gauss2d/utils.py

```diff
@@ -0,0 +1,179 @@
+# This file is part of gauss2d.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import numpy as np
+from _gauss2d import loglike_gaussians_pixel as loglike_gaussians_pixel_pb
+
+
+def covar_to_ellipse(sigma_x_sq, sigma_y_sq, cov_xy, degrees=False):
+    """ Convert covariance matrix terms to ellipse major axis, axis ratio and position angle representation.
+
+    Parameters
+    ----------
+    sigma_x_sq, sigma_y_sq : `float` or array-like
+        x- and y-axis squared standard deviations of a 2-dimensional normal distribution (diagonal terms of
+        its covariance matrix). Must be scalar or identical length array-likes.
+    cov_xy : `float` or array-like
+        x-y covariance of a of a 2-dimensional normal distribution (off-diagonal term of its covariance
+        matrix). Must be scalar or identical length array-likes.
+    degrees : `bool`
+        Whether to return the position angle in degrees instead of radians.
+
+    Returns
+    -------
+    r_major, axrat, angle : `float` or array-like
+        Converted major-axis radius, axis ratio and position angle (counter-clockwise from the +x axis) of
+        the ellipse defined by each set of input covariance matrix terms.
+
+    Notes
+    -----
+    The eigenvalues from the determinant of a covariance matrix are:
+    |a-m b|
+    |b c-m|
+    det = (a-m)(c-m) - b^2 = ac - (a+c)m + m^2 - b^2 = m^2 - (a+c)m + (ac-b^2)
+    m = ((a+c) +/- sqrt((a+c)^2 - 4(ac-b^2)))/2
+
+    The values from this function should match those from `Ellipse.make_ellipse_major` to within rounding
+    error, except in the special case of sigma_x == sigma_y == 0, which returns a NaN axis ratio here by
+    default. It is mainly intended to be more convenient for array inputs.
+    """
+    apc = sigma_x_sq + sigma_y_sq
+    x = apc/2
+    pm = np.sqrt(apc**2 - 4*(sigma_x_sq*sigma_y_sq - cov_xy**2))/2
+
+    r_major = x + pm
+    axrat = np.sqrt((x - pm)/r_major)
+    r_major = np.sqrt(r_major)
```

Comment:
I did it because I think it would be better for floating point precision purposes to divide before the sqrt, although I could be wrong. It probably won't matter much unless the code is later templated to use {{float}} instead of {{double}}.